File: Sources/DoNotTypeCore/ProviderProbe.swift
import Foundation

/// One cheap round trip that answers "will this key work when it matters?".
///
/// It exists because the alternative is finding out mid-dictation: the key is only read when the
/// user has already spoken, so a wrong or absent one surfaced as a failed recording rather than as
/// a setting that needed fixing. Asking at launch and after every edit turns that into a question
/// answered while nothing is at stake.
///
/// The probe deliberately distinguishes *rejected* from *no answer*. A 401 means the user has to do
/// something; a timeout on a train means nothing at all, and reporting the second as the first
/// would train people to ignore both.
///
/// It answers a second question as well: whether the audio arrives. Every probe is a recording, so
/// an endpoint that cannot take one — a text-only relay, a `vllm serve` in front of a text-only
/// checkpoint — is rejected here rather than discovered mid-dictation. The limit worth knowing is
/// that a provider reporting no usage at all still passes: `assertAudioWasProcessed` can only call
/// a *reported* zero a dropped recording, and silence about it stays unprovable either way.
public enum ProviderProbe {
    public enum Outcome: Sendable, Equatable {
        /// The provider answered and the key worked.
        case accepted
        /// The provider answered and something about this configuration has to change.
        case rejected(String)
        /// No usable answer — offline, timed out, or the provider itself is unwell. Says nothing
        /// about the key.
        case inconclusive(String)
    }

    /// - Parameter model: only used by model backends; recognition endpoints ignore it.
    public static func check(
        _ provider: any TranscriptionProvider, model: String
    ) async -> Outcome {
        // Audio, for every backend. Recognition endpoints always got silence, because they reject
        // a text-only request by design; model backends got text, and that turned out to be the
        // one thing this check could not afford to skip. "OpenAI-compatible" is a claim about the
        // request shape, so a relay or a text-only checkpoint answers the text probe perfectly and
        // then has nowhere to put a recording — which the user found out mid-dictation, from a
        // transcript the model invented. Probing with the same shape a dictation uses moves that
        // discovery to the settings window, where it costs nothing.
        //
        // A quarter second is deliberate rather than merely cheap: it is
        // `SpeechActivity.minimumSpeechMilliseconds`, the shortest audio this app will ever send
        // for real. Anything a provider does to this clip — including billing zero audio tokens
        // for it, which is how a dropped recording is detected — it would do to a real dictation.
        let parts: [InputPart] = [.audio(data: silentProbeWAV, mimeType: "audio/wav")]

        do {
            _ = try await provider.transcribe(
                TranscriptionRequest(
                    model: model,
                    systemInstruction: "You are a transcription engine.",
                    parts: parts))
            return .accepted
        } catch ProviderError.emptyOutput {
            // Silence transcribes to nothing, which is the correct answer and proves the round
            // trip worked.
            return .accepted
        } catch {
            // The same triage the dictation path uses, so a key check and a failed dictation never
            // disagree about whether the user has to act. `isOnline: true` because being offline
            // is exactly the case this must report as inconclusive rather than as a bad key.
            let advice = FailureAdvice.describe(error, isOnline: true)

            // `needsUserAction` alone is not the whole question here. A plain 4xx is deliberately
            // *not* the user's to act on during a dictation — nothing in Settings fixes a request
            // this app built wrong — so it is triaged as "worth reporting". The probe builds a
            // fixed, minimal request, and that reasoning inverts: if this one is refused, what is
            // wrong is the configuration it was sent with, and the endpoint field is where it gets
            // fixed. A text-only relay answering "input_audio is not supported" arrives exactly
            // this way, and reporting it as "could not ask" would hide the one answer the check
            // exists to give. Retryability is the honest discriminator — advice that says retrying
            // will not help is advice about a setting, not about the network.
            return advice.needsUserAction || !advice.isRetryable
                ? .rejected(advice.message)
                : .inconclusive(error.localizedDescription)
        }
    }

    /// A quarter-second of 16 kHz mono silence, built rather than shipped as a fixture so the
    /// bundle does not carry a resource used by one check.
    static let silentProbeWAV: Data = {
        let sampleRate = 16_000
        let samples = sampleRate / 4
        let dataBytes = samples * 2

        var wav = Data()
        func append(_ string: String) { wav.append(Data(string.utf8)) }
        func append(u32 value: UInt32) {
            withUnsafeBytes(of: value.littleEndian) { wav.append(contentsOf: $0) }
        }
        func append(u16 value: UInt16) {
            withUnsafeBytes(of: value.littleEndian) { wav.append(contentsOf: $0) }
        }

        append("RIFF")
        append(u32: UInt32(36 + dataBytes))
        append("WAVEfmt ")
        append(u32: 16)                                  // PCM header length
        append(u16: 1)                                   // PCM
        append(u16: 1)                                   // mono
        append(u32: UInt32(sampleRate))
        append(u32: UInt32(sampleRate * 2))              // byte rate
        append(u16: 2)                                   // block align
        append(u16: 16)                                  // bits per sample
        append("data")
        append(u32: UInt32(dataBytes))
        wav.append(Data(repeating: 0, count: dataBytes))
        return wav
    }()
}

File: android/app/src/main/kotlin/app/donottype/core/ProviderProbe.kt
package app.donottype.core

import android.os.SystemClock
import java.io.ByteArrayOutputStream
import java.util.Locale
import kotlin.math.roundToInt

/**
 * One cheap round trip that answers "will this key work when it matters?".
 *
 * It exists because the alternative is finding out mid-dictation: the key is only read once the
 * user has already spoken, so a wrong or absent one surfaced as a failed recording rather than as
 * a setting that needed fixing. Asking while nothing is at stake turns that into a question
 * answered before the first word.
 *
 * Shared rather than copied because it now has two callers -- the settings screen and onboarding.
 * A second copy would be a second opinion about whether a key works, and the one place a user
 * checks that is the one place it must not disagree with itself.
 *
 * Mirrors `ProviderProbe.check` in the Swift core. The one deliberate difference is what comes
 * back: Swift returns an outcome its view formats, this returns the line to show, because both
 * Android callers show the same sentence and duplicating the formatting is how they would drift.
 */
object ProviderProbe {

    /**
     * @param accepted whether the provider answered and took the key. Drives the colour; the two
     *   ways of failing (refused, unreachable) both read as bad news and both say why in [message].
     */
    data class Result(val message: String, val accepted: Boolean)

    /**
     * A quarter second of silence sent the way a dictation is sent.
     *
     * Audio for every backend, including the model ones. They used to get a text round trip, which
     * a text-only relay or checkpoint answers perfectly well before dropping the first real
     * recording -- "OpenAI-compatible" is a claim about the request shape, not about whether there
     * is anywhere to put a recording. Probing with the shape a dictation uses moves that discovery
     * here, where it costs nothing.
     */
    suspend fun check(provider: ProviderKind, apiKey: String, model: String): Result {
        val client = ProviderFactory.create(provider, apiKey, model)
        val parts = listOf(InputPart.Audio(silentProbeWav(), "audio/wav"))

        val started = SystemClock.elapsedRealtimeNanos()
        val outcome = runCatching {
            client.transcribe("You are a transcription engine.", parts)
        }
        val latency = latencyLabel(SystemClock.elapsedRealtimeNanos() - started)

        return outcome.fold(
            onSuccess = { Result("✓ Reachable, key accepted · $latency", accepted = true) },
            onFailure = { error ->
                // Silence transcribes to nothing, which is the correct answer and proves the round
                // trip worked.
                if (error.message?.contains("no output", ignoreCase = true) == true) {
                    Result("✓ Reachable, key accepted · $latency", accepted = true)
                } else {
                    // Never shortened. What the provider said is the whole of what the user has to
                    // act on, and it is the line they will paste into a bug report.
                    Result("✗ ${error.message} · $latency", accepted = false)
                }
            },
        )
    }

    /** How long the round trip took, in the unit that reads at that scale. */
    fun latencyLabel(nanoseconds: Long): String {
        val milliseconds = nanoseconds.coerceAtLeast(0) / 1_000_000.0
        return if (milliseconds < 1_000) {
            "${milliseconds.roundToInt()} ms"
        } else {
            String.format(Locale.ROOT, "%.2f s", milliseconds / 1_000)
        }
    }

    /**
     * A quarter-second of 16 kHz mono silence, built rather than shipped as an asset so the APK
     * does not carry a resource used by one button.
     *
     * The duration is deliberate rather than merely cheap: it is the shortest audio this app will
     * ever send for real, so anything a provider does to this clip it would do to a dictation.
     */
    fun silentProbeWav(): ByteArray {
        val sampleRate = 16_000
        val dataBytes = sampleRate / 4 * 2
        val out = ByteArrayOutputStream()
        fun ascii(value: String) = out.write(value.toByteArray(Charsets.US_ASCII))
        fun u32(value: Int) = out.write(
            byteArrayOf(
                value.toByte(),
                (value shr 8).toByte(),
                (value shr 16).toByte(),
                (value shr 24).toByte(),
            ),
        )
        fun u16(value: Int) = out.write(byteArrayOf(value.toByte(), (value shr 8).toByte()))

        ascii("RIFF"); u32(36 + dataBytes); ascii("WAVEfmt ")
        u32(16); u16(1); u16(1); u32(sampleRate); u32(sampleRate * 2); u16(2); u16(16)
        ascii("data"); u32(dataBytes)
        out.write(ByteArray(dataBytes))
        return out.toByteArray()
    }
}
